### ISSUE TITLE
Use Streaming Extractor Music when only File Path Descriptor is available

### ISSUE THREAD

**Issue by amCap1712** [MTG/essentia#1005]
Hi! I am trying to write an AcousticBrainz client for Android. I know that AcousticBrainz uses [Streaming Extractor Music](https://github.com/MTG/essentia/blob/master/src/examples/streaming_extractor_music.cpp) to generate the relevant data. It requires the file path of the input file. However, on Android it is only feasible to obtain the `FileDescriptor` or `InputStream` for the audio file. How can I modify this example to use a file descriptor instead of the file name instead ?

**Comment by dbogdanov** [comment]
The extractor is using `MusicExtractor` algorithm which relies on `AudioLoader` and `avformat_open_input`therein.

One related task we have on our wishlist is to separate MusicExtractor functionality in two algorithms: one that loads audio and another one that works with the loaded audio stream to compute descriptors. In this case it will be possible to use the extractor together with any other audio loader.

Btw, we've done some some work on build scripts for the dependencies on Android, if that's relevant: 
https://github.com/MTG/essentia/commits/build-android

**Comment by amCap1712** [comment]
Regarding, the android build scripts I had myself done some work on that and opened a PR. I was intending to do more work but it seems it is already covered in the branch you mentioned.

**Comment by dbogdanov** [comment]
Just need to mention that the android branch is not finished - there was still some issue when running `streaming streaming_music_extractor` from the command line on Android, due to one of the dependencies. Haven't look further into that yet.

**Comment by amCap1712** [comment]
@dbogdanov Do you some logcat or steps to reproduce the issue ? I can try to look into it.

**Comment by amCap1712** [comment]
@dbogdanov I worked around the File issue and was able to use streaming_extractor_music. The app does not crash but I get result code 1. Anyway to enable logging for the to get further details to debug the issue ?

**Comment by dbogdanov** [comment]
You can [configure the logging level](https://essentia.upf.edu/design_overview.html#logging-in-c) [somewhere here](https://github.com/MTG/essentia/blob/master/src/examples/streaming_extractor_music.cpp#L52).

For example:
```
 setDebugLevel(EAll);
```

**Comment by amCap1712** [comment]
@dbogdanov I used the scripts to build dependencies for android and found some issues. I'll try to open a PR fixing them. I had a question, can ffmpeg be bumped up to 3.x or 4.x ?

**Comment by dbogdanov** [comment]
We cannot update the default FFmpeg version for all builds at the moment for consistency, but you can overwrite the version [specifically for Android here](https://github.com/MTG/essentia/blob/build-android/packaging/build_config_android.sh#L61).

**Comment by amCap1712** [comment]
Yes, I understand that. I meant that is it safe to upgrade as there are breaking changes from 2.8 to 3.x and 4.x ?

**Comment by amCap1712** [comment]
Hi! I am able to build and run essentia on Android now. How much time should I expect the algorithm to run on an mp3 without passing any custom profile? Android discards all outputs to stdio and stderr so there is no way of telling what is going under the hood. There is an option to redirect these outputs to logcat but only for rooted devices. I'll try this on an emulator later.

**Comment by amCap1712** [comment]
Update: It took approx 15 mins to execute but it worked.

**Comment by dbogdanov** [comment]
15 mins is really slow and unusable. It takes ony 15-30 seconds for a 5 min track on my machines (laptop, pc).

**Comment by dbogdanov** [comment]
You can try running a much simpler extractor, for example `essentia_standard_mfcc` to see if it is also very slow.

**Comment by amCap1712** [comment]
@dbogdanov I ran `essentia_standard_mfcc`and it completed instantaneously within a few seconds. Can I do something to help debug the cause of this slowness ? Also, I am using a statically built essentia (to avoid the linking errors I was facing at runtime), could this be a problem ?

**Comment by amCap1712** [comment]
@dbogdanov I tested `streaming_mfcc` and `streamng_extractor_la-capula` which taken less than  1 minute and 2-3 minutes to execute respectively.

**Comment by amCap1712** [comment]
@dbogdanov I have made a java wrapper for the Streaming Extractor Music on Android. https://github.com/metabrainz/acousticbrainz-android in case it helps in debugging the issue of slowness.
